src/symbols_macos.cpp

```diff
@@ -16,14 +16,127 @@
 
 #ifdef __APPLE__
 
+#include <mach-o/loader.h>
+#include <cstring>
+#include <mach-o/nlist.h>
 #include "symbols.h"
+#include <sys/mman.h>
+#include <mach-o/dyld_images.h>
+#include <fcntl.h>
 
+// Workaround for newer macosx versions: since 10.12.x _dyld_get_all_image_infos exists, but is not exported
+extern "C" const struct dyld_all_image_infos* _dyld_get_all_image_infos();
+
+class MachOParser {
+  private:
+    void* _lib_addr;
+
+    static load_command* find_command(mach_header_64* header, uint32_t command) {
+        load_command* result = (load_command*)((uint64_t)header + sizeof(mach_header_64));
+
+        if (command == 0) {
+            return result;
+        }
+
+        for (uint32_t i = 0; i < header->ncmds; i++) {
+            if (result->cmd == command) {
+                break;
+            }
+
+            result = (load_command*)((uint64_t)result + result->cmdsize);
+        }
+        return result;
+    }
+
+  public:
+    MachOParser(void* _lib_addr) : _lib_addr(_lib_addr) {
+    }
+
+    uintptr_t find_symbol_offset(const char* symbol_name) {
+        load_command* command = find_command((mach_header_64*)_lib_addr, LC_SYMTAB);
+        symtab_command* symtab = (symtab_command*)command;
+        nlist_64* symbol_table = (nlist_64*)((char*)_lib_addr + symtab->symoff);
+
+        const char* str_table = (char*)_lib_addr + symtab->stroff;
+        // Scan through whole symbol table
+        for (uint32_t sym_n = 0; sym_n < symtab->nsyms; ++sym_n) {
+            uint32_t str_offset = symbol_table[sym_n].n_un.n_strx;
+            const char* sym_name = &str_table[str_offset];
+            uintptr_t offset = symbol_table[sym_n].n_value;
+            if (strcmp(sym_name, symbol_name) == 0 && offset != 0) {
+                return offset;
+            }
+        }
+        return 0;
+    }
+};
+
+void add_symbol(MachOParser &parser, uintptr_t lib_addr, const char* symbol_name, NativeCodeCache* cc) {
+    uintptr_t offset = parser.find_symbol_offset(symbol_name);
+    cc->add((void*)(lib_addr + offset), sizeof(uintptr_t), symbol_name + 1);
+}
 
 void Symbols::parseKernelSymbols(NativeCodeCache* cc) {
 }
 
 int Symbols::parseMaps(NativeCodeCache** array, int size) {
-    return 0;
+
+    const dyld_all_image_infos* dyld_all_image_infos = _dyld_get_all_image_infos();
+    const char* lib_path = NULL;
+    uintptr_t loaded_lib_addr = 0;
+    for (int i = 0; i < dyld_all_image_infos->infoArrayCount; i++) {
+        const char* path = dyld_all_image_infos->infoArray[i].imageFilePath;
+        size_t length = strlen(path);
+        if (strcmp(path + length - 12, "libjvm.dylib") == 0) {
+            lib_path = path;
+            loaded_lib_addr = (uintptr_t)dyld_all_image_infos->infoArray[i].imageLoadAddress;
+        }
+    }
```

Comment:
I did it intentionally: I'm trying to avoid work which is not really necessary.
It's not a big deal to add all symbols when native call stacks will be supported

Reply:
Yes, this will make the code consistent across platforms and will facilitate further support for native stack traces on Mac. At least we may start with parsing only libjvm symbols as you suggested.